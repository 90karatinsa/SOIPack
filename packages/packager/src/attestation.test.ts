import { createHash, X509Certificate, createPublicKey, verify as verifySignature } from 'crypto';
import { readFileSync } from 'fs';
import path from 'path';

import type { LedgerAwareManifest } from './index';
import { generateAttestation, serializeAttestationDocument } from './attestation';
import type { ManifestSignatureBundle } from './index';

const DEV_CERT_BUNDLE_PATH = path.resolve(__dirname, '../../../test/certs/dev.pem');

const loadDevCredentials = (): { privateKeyPem: string; certificatePem: string; bundlePem: string } => {
  const bundlePem = readFileSync(DEV_CERT_BUNDLE_PATH, 'utf8');
  const certificateMatch = bundlePem.match(/-----BEGIN CERTIFICATE-----[\s\S]+?-----END CERTIFICATE-----/);
  if (!certificateMatch) {
    throw new Error('Dev sertifikası bulunamadı.');
  }

  return {
    privateKeyPem: bundlePem,
    certificatePem: certificateMatch[0],
    bundlePem,
  };
};

describe('Attestation', () => {
  const manifest: LedgerAwareManifest = {
    createdAt: '2024-02-01T10:15:00.000Z',
    toolVersion: '0.2.0',
    files: [
      {
        path: 'reports/summary.txt',
        sha256: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
      },
      {
        path: 'evidence/log.csv',
        sha256: 'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb',
      },
    ],
    stage: 'SOI-1',
    ledger: {
      root: 'cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc',
      previousRoot: 'dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd',
    },
    merkle: {
      algorithm: 'ledger-merkle-v1',
      root: 'eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
      manifestDigest: 'ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
      snapshotId: 'manifest:SOI-1:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
    },
    sbom: {
      path: 'sbom.spdx.json',
      algorithm: 'sha256',
      digest: '1111111111111111111111111111111111111111111111111111111111111111',
    },
  };

  const manifestDigest = '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef';

  it('Attestation Packager attestation generates deterministic provenance statements and Ed25519 JWS signatures', async () => {
    const { bundlePem, certificatePem } = loadDevCredentials();
    const certificate = new X509Certificate(certificatePem);
    const publicKeyPem = certificate.publicKey.export({ format: 'pem', type: 'spki' }).toString();

    const signatureBundle: ManifestSignatureBundle = {
      signature: 'stub-signature',
      certificate: certificatePem,
      manifestDigest: { algorithm: 'SHA-256', hash: manifestDigest },
      ledgerRoot: manifest.ledger?.root ?? null,
      previousLedgerRoot: manifest.ledger?.previousRoot ?? null,
    };

    const result = await generateAttestation({
      manifest,
      manifestDigest,
      sbom: manifest.sbom!,
      files: manifest.files.map((file) => ({ path: file.path, sha256: file.sha256 })),
      packageName: 'soi-pack-test.zip',
      manifestSignature: signatureBundle,
      builderId: 'https://soipack.dev/test-builder',
      invocationId: 'test-run-123',
      signing: {
        privateKeyPem: bundlePem,
        publicKeyPem,
        keyId: 'test-dev-key',
      },
      ledgerDeltaSummary: {
        currentRoot: manifest.ledger?.root ?? null,
        previousRoot: manifest.ledger?.previousRoot ?? null,
        objectiveIds: ['OBJ-1', 'OBJ-2'],
      },
    });

    const serialized = serializeAttestationDocument(result.document);
    expect(serialized).toMatchSnapshot('attestation-document');

    const payloadJson = JSON.parse(result.payload) as {
      predicate: { runDetails: { metadata: { ledgerDelta: unknown; signature: { ledgerDelta: unknown } } } };
    };

    expect(payloadJson.predicate.runDetails.metadata.ledgerDelta).toEqual({
      currentRoot: manifest.ledger?.root ?? null,
      previousRoot: manifest.ledger?.previousRoot ?? null,
      objectiveIds: ['OBJ-1', 'OBJ-2'],
    });
    expect(payloadJson.predicate.runDetails.metadata.signature.ledgerDelta).toEqual(
      payloadJson.predicate.runDetails.metadata.ledgerDelta,
    );

    const [encodedHeader, encodedPayload, encodedSignature] = result.signature.jws.split('.');
    expect(encodedHeader).toBe(result.signature.protected);
    expect(encodedPayload).toBe(Buffer.from(result.payload, 'utf8').toString('base64url'));
    expect(encodedSignature).toBe(result.signature.signature);

    const verificationPayload = Buffer.from(`${encodedHeader}.${encodedPayload}`, 'utf8');
    const verified = verifySignature(
      null,
      verificationPayload,
      createPublicKey(result.signature.publicKey),
      Buffer.from(encodedSignature, 'base64url'),
    );
    expect(verified).toBe(true);

    const expectedDigest = createHash('sha256').update(result.payload, 'utf8').digest('hex');
    expect(result.document.statementDigest).toEqual({ algorithm: 'sha256', digest: expectedDigest });
  });

  it('includes ledger delta summaries and hybrid signature metadata for post-quantum bundles', async () => {
    const { bundlePem, certificatePem } = loadDevCredentials();
    const certificate = new X509Certificate(certificatePem);
    const publicKeyPem = certificate.publicKey.export({ format: 'pem', type: 'spki' }).toString();

    const pqSignature = Buffer.from('post-quantum-signature');
    const pqSignatureEncoded = pqSignature.toString('base64url');

    const overrideLedgerDelta = {
      currentRoot: '9999999999999999999999999999999999999999999999999999999999999999',
      previousRoot: '8888888888888888888888888888888888888888888888888888888888888888',
      objectiveIds: ['OBJ-3', 'OBJ-1', 'OBJ-2', 'OBJ-1', ''],
    };

    const signatureBundle: ManifestSignatureBundle = {
      signature: 'hybrid-stub-signature',
      certificate: certificatePem,
      manifestDigest: { algorithm: 'SHA-256', hash: manifestDigest },
      ledgerRoot: manifest.ledger?.root ?? null,
      previousLedgerRoot: manifest.ledger?.previousRoot ?? null,
      postQuantumSignature: {
        algorithm: 'SPHINCS+-SHA2-128s',
        signature: pqSignatureEncoded,
        publicKey: 'pq-public-key-material',
      },
    };

    const result = await generateAttestation({
      manifest,
      manifestDigest,
      sbom: manifest.sbom!,
      files: manifest.files.map((file) => ({ path: file.path, sha256: file.sha256 })),
      packageName: 'soi-pack-test.zip',
      manifestSignature: signatureBundle,
      signing: {
        privateKeyPem: bundlePem,
        publicKeyPem,
      },
      ledgerDeltaSummary: overrideLedgerDelta,
    });

    const payloadJson = JSON.parse(result.payload) as {
      predicate: {
        runDetails: {
          metadata: {
            ledgerDelta: { currentRoot: string | null; previousRoot: string | null; objectiveIds: string[] };
            signature: {
              ledgerDelta: { currentRoot: string | null; previousRoot: string | null; objectiveIds: string[] };
              postQuantum?: { algorithm: string; signature: string; publicKey: string };
            };
          };
        };
      };
    };

    expect(payloadJson.predicate.runDetails.metadata.ledgerDelta).toEqual({
      currentRoot: overrideLedgerDelta.currentRoot,
      previousRoot: overrideLedgerDelta.previousRoot,
      objectiveIds: ['OBJ-1', 'OBJ-2', 'OBJ-3'],
    });
    expect(payloadJson.predicate.runDetails.metadata.signature.ledgerDelta).toEqual(
      payloadJson.predicate.runDetails.metadata.ledgerDelta,
    );
    expect(payloadJson.predicate.runDetails.metadata.signature.postQuantum).toEqual({
      algorithm: 'SPHINCS+-SHA2-128s',
      signature: pqSignatureEncoded,
      publicKey: 'pq-public-key-material',
    });

    const [encodedHeader, encodedPayload, encodedSignature, encodedPostQuantum] = result.signature.jws.split('.');
    expect(encodedHeader).toBe(result.signature.protected);
    expect(encodedPayload).toBe(Buffer.from(result.payload, 'utf8').toString('base64url'));
    expect(encodedSignature).toBe(result.signature.signature);
    expect(encodedPostQuantum).toBeUndefined();

    const verificationPayload = Buffer.from(`${encodedHeader}.${encodedPayload}`, 'utf8');
    const verified = verifySignature(
      null,
      verificationPayload,
      createPublicKey(result.signature.publicKey),
      Buffer.from(encodedSignature, 'base64url'),
    );
    expect(verified).toBe(true);
  });
});
