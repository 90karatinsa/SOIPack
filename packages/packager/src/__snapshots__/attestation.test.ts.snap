// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attestation Attestation Packager attestation generates deterministic provenance statements and Ed25519 JWS signatures: attestation-document 1`] = `
"{
  "signatures": [
    {
      "algorithm": "EdDSA",
      "jws": "eyJhbGciOiJFZERTQSIsImtpZCI6InRlc3QtZGV2LWtleSIsInR5cCI6ImFwcGxpY2F0aW9uL3ZuZC5pbi10b3RvK2pzb24ifQ.eyJfdHlwZSI6Imh0dHBzOi8vaW4tdG90by5pby9TdGF0ZW1lbnQvdjAuMSIsInByZWRpY2F0ZSI6eyJidWlsZERlZmluaXRpb24iOnsiYnVpbGRUeXBlIjoiaHR0cHM6Ly9zb2lwYWNrLmRldi9hdHRlc3RhdGlvbnMvcGFja2FnZXIvdjEiLCJleHRlcm5hbFBhcmFtZXRlcnMiOnsibGVkZ2VyIjp7InByZXZpb3VzUm9vdCI6ImRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGQiLCJyb290IjoiY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjYyJ9LCJwYWNrYWdlTmFtZSI6InNvaS1wYWNrLXRlc3QuemlwIiwic2JvbSI6eyJhbGdvcml0aG0iOiJzaGEyNTYiLCJkaWdlc3QiOiIxMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExIiwicGF0aCI6InNib20uc3BkeC5qc29uIn0sInN0YWdlIjoiU09JLTEifSwicmVzb2x2ZWREZXBlbmRlbmNpZXMiOlt7ImRpZ2VzdCI6eyJzaGEyNTYiOiJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiYmJiIn0sInVyaSI6ImZpbGU6ZXZpZGVuY2UvbG9nLmNzdiJ9LHsiZGlnZXN0Ijp7InNoYTI1NiI6ImFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWEifSwidXJpIjoiZmlsZTpyZXBvcnRzL3N1bW1hcnkudHh0In1dfSwicnVuRGV0YWlscyI6eyJidWlsZGVyIjp7ImlkIjoiaHR0cHM6Ly9zb2lwYWNrLmRldi90ZXN0LWJ1aWxkZXIifSwiYnlwcm9kdWN0cyI6W3siZGlnZXN0Ijp7InNoYTI1NiI6ImY1NDA0Y2FhYTc4OTExNWMyODdlNzVmMzAyNjQwZmZkNzQwMmFhZjllODQwZWQ4ODE5ZjIyZTYxZGM5MWYzMzEifSwibWVkaWFUeXBlIjoiYXBwbGljYXRpb24vam9zZStqc29uIiwibmFtZSI6Im1hbmlmZXN0LXNpZ25hdHVyZSIsInVyaSI6InVybjpzb2lwYWNrOm1hbmlmZXN0LXNpZ25hdHVyZSJ9XSwibWV0YWRhdGEiOnsiaW5wdXRzRGlnZXN0IjoiZjU1YWFiYzQ4ZTkxNDY5MDA2ODZiMWNmNjAwYWRjMTE4YTM1ZDJiODUwODliN2ZhYTFjZTNlZDdiNWFlMTFhNSIsImludm9jYXRpb25JZCI6InRlc3QtcnVuLTEyMyIsImxlZGdlckRlbHRhIjp7ImN1cnJlbnRSb290IjoiY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjYyIsIm9iamVjdGl2ZUlkcyI6WyJPQkotMSIsIk9CSi0yIl0sInByZXZpb3VzUm9vdCI6ImRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGQifSwibGVkZ2VyUm9vdCI6ImNjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2MiLCJtYW5pZmVzdENyZWF0ZWRBdCI6IjIwMjQtMDItMDFUMTA6MTU6MDAuMDAwWiIsIm1hbmlmZXN0RGlnZXN0IjoiMDEyMzQ1Njc4OWFiY2RlZjAxMjM0NTY3ODlhYmNkZWYwMTIzNDU2Nzg5YWJjZGVmMDEyMzQ1Njc4OWFiY2RlZiIsIm1hbmlmZXN0U3RhZ2UiOiJTT0ktMSIsIm1lcmtsZVJvb3QiOiJlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlZWVlIiwibWVya2xlU25hcHNob3QiOiJtYW5pZmVzdDpTT0ktMTpmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmZmIiwicGFja2FnZU5hbWUiOiJzb2ktcGFjay10ZXN0LnppcCIsInByZXZpb3VzTGVkZ2VyUm9vdCI6ImRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGQiLCJzYm9tRGlnZXN0IjoiMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMSIsInNpZ25hdHVyZSI6eyJjZXJ0aWZpY2F0ZVNoYTI1NiI6IjhjMDVmZmYyYTZlMjdkNjY0YTljNmZlZmExNzdlOGRlNmZiZWFkNjg4MDZkOTNiOTkyYzU1NDVlYzdiMjAxMjAiLCJkaWdlc3QiOiIwMTIzNDU2Nzg5YWJjZGVmMDEyMzQ1Njc4OWFiY2RlZjAxMjM0NTY3ODlhYmNkZWYwMTIzNDU2Nzg5YWJjZGVmIiwibGVkZ2VyRGVsdGEiOnsiY3VycmVudFJvb3QiOiJjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjIiwib2JqZWN0aXZlSWRzIjpbIk9CSi0xIiwiT0JKLTIiXSwicHJldmlvdXNSb290IjoiZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZCJ9LCJsZWRnZXJSb290IjoiY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjYyIsInByZXZpb3VzTGVkZ2VyUm9vdCI6ImRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGRkZGQifSwidG9vbFZlcnNpb24iOiIwLjIuMCJ9fX0sInByZWRpY2F0ZVR5cGUiOiJodHRwczovL3Nsc2EuZGV2L3Byb3ZlbmFuY2UvdjEiLCJzdWJqZWN0IjpbeyJkaWdlc3QiOnsic2hhMjU2IjoiMDEyMzQ1Njc4OWFiY2RlZjAxMjM0NTY3ODlhYmNkZWYwMTIzNDU2Nzg5YWJjZGVmMDEyMzQ1Njc4OWFiY2RlZiJ9LCJuYW1lIjoibWFuaWZlc3QuanNvbiJ9LHsiZGlnZXN0Ijp7InNoYTI1NiI6ImY1NDA0Y2FhYTc4OTExNWMyODdlNzVmMzAyNjQwZmZkNzQwMmFhZjllODQwZWQ4ODE5ZjIyZTYxZGM5MWYzMzEifSwibmFtZSI6Im1hbmlmZXN0LnNpZyJ9LHsiZGlnZXN0Ijp7InNoYTI1NiI6IjExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTEifSwibmFtZSI6InNib20uc3BkeC5qc29uIn1dfQ.4eQ2ne2vSSdef50CkA2SFPFro2J78dWq8RIJ9S9hgBtnTQA7o8vBeI6HgaRKuP6eueCYY9evvhzpz-nZdQ7pCA",
      "keyId": "test-dev-key",
      "protected": "eyJhbGciOiJFZERTQSIsImtpZCI6InRlc3QtZGV2LWtleSIsInR5cCI6ImFwcGxpY2F0aW9uL3ZuZC5pbi10b3RvK2pzb24ifQ",
      "protectedHeader": {
        "alg": "EdDSA",
        "kid": "test-dev-key",
        "typ": "application/vnd.in-toto+json"
      },
      "publicKey": "-----BEGIN PUBLIC KEY-----\\nMCowBQYDK2VwAyEAfbDiOb9ZkQEBEWRrVcnwS1iusCtSDdp/CTBibt2H1ts=\\n-----END PUBLIC KEY-----\\n",
      "signature": "4eQ2ne2vSSdef50CkA2SFPFro2J78dWq8RIJ9S9hgBtnTQA7o8vBeI6HgaRKuP6eueCYY9evvhzpz-nZdQ7pCA"
    }
  ],
  "statement": {
    "_type": "https://in-toto.io/Statement/v0.1",
    "predicate": {
      "buildDefinition": {
        "buildType": "https://soipack.dev/attestations/packager/v1",
        "externalParameters": {
          "ledger": {
            "previousRoot": "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd",
            "root": "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"
          },
          "packageName": "soi-pack-test.zip",
          "sbom": {
            "algorithm": "sha256",
            "digest": "1111111111111111111111111111111111111111111111111111111111111111",
            "path": "sbom.spdx.json"
          },
          "stage": "SOI-1"
        },
        "resolvedDependencies": [
          {
            "digest": {
              "sha256": "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
            },
            "uri": "file:evidence/log.csv"
          },
          {
            "digest": {
              "sha256": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
            },
            "uri": "file:reports/summary.txt"
          }
        ]
      },
      "runDetails": {
        "builder": {
          "id": "https://soipack.dev/test-builder"
        },
        "byproducts": [
          {
            "digest": {
              "sha256": "f5404caaa789115c287e75f302640ffd7402aaf9e840ed8819f22e61dc91f331"
            },
            "mediaType": "application/jose+json",
            "name": "manifest-signature",
            "uri": "urn:soipack:manifest-signature"
          }
        ],
        "metadata": {
          "inputsDigest": "f55aabc48e9146900686b1cf600adc118a35d2b85089b7faa1ce3ed7b5ae11a5",
          "invocationId": "test-run-123",
          "ledgerDelta": {
            "currentRoot": "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
            "objectiveIds": [
              "OBJ-1",
              "OBJ-2"
            ],
            "previousRoot": "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd"
          },
          "ledgerRoot": "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
          "manifestCreatedAt": "2024-02-01T10:15:00.000Z",
          "manifestDigest": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
          "manifestStage": "SOI-1",
          "merkleRoot": "eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "merkleSnapshot": "manifest:SOI-1:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
          "packageName": "soi-pack-test.zip",
          "previousLedgerRoot": "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd",
          "sbomDigest": "1111111111111111111111111111111111111111111111111111111111111111",
          "signature": {
            "certificateSha256": "8c05fff2a6e27d664a9c6fefa177e8de6fbead68806d93b992c5545ec7b20120",
            "digest": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
            "ledgerDelta": {
              "currentRoot": "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
              "objectiveIds": [
                "OBJ-1",
                "OBJ-2"
              ],
              "previousRoot": "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd"
            },
            "ledgerRoot": "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
            "previousLedgerRoot": "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd"
          },
          "toolVersion": "0.2.0"
        }
      }
    },
    "predicateType": "https://slsa.dev/provenance/v1",
    "subject": [
      {
        "digest": {
          "sha256": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
        },
        "name": "manifest.json"
      },
      {
        "digest": {
          "sha256": "f5404caaa789115c287e75f302640ffd7402aaf9e840ed8819f22e61dc91f331"
        },
        "name": "manifest.sig"
      },
      {
        "digest": {
          "sha256": "1111111111111111111111111111111111111111111111111111111111111111"
        },
        "name": "sbom.spdx.json"
      }
    ]
  },
  "statementDigest": {
    "algorithm": "sha256",
    "digest": "d42b17abc33a0ec7f9de4d94fcbe8cea8cdfa5f9734d392357a0e7c405b53cfe"
  }
}"
`;
